### DISCUSSION TITLE
[Question] Routing HTTPS requests to HTTPS OR HTTP if no port 443 is available

### DISCUSSION THREAD

**Discussion by cnaslain** [nginx-proxy/nginx-proxy#1661] [Q&A]
Hi,
I have a question...
I've setup nginx-proxy with several apps running in the same service using docker-compose.
Everything works fine using only VIRTUAL_HOST environment parameters:
- Internet -> app1.example.com:80 --> nginx_proxy --> docker app1:80
- Internet -> app2.example.com:80 --> nginx_proxy --> docker app2:80
- Internet -> app3.example.com:80 --> nginx_proxy --> docker app1:80
- Internet -> app1.example.com:443 --> nginx_proxy --> docker app1:443
- Internet -> app2.example.com:443 --> nginx_proxy --> docker app2:443
- Internet -> app3.example.com:443 --> nginx_proxy --> NOT OPEN on the container

My question is: If I configure the nginx-proxy with only ports 443:443 (no 80 exposed); can I configure the apps to expose their normal 80 ports to the 443 so that all requests from the proxy using 443 will be redirected to the port 80 of the app?
- Internet -> app1.example.com:80 --> nginx_proxy --> NOT OPEN on the proxy
- Internet -> app2.example.com:80 --> nginx_proxy --> NOT OPEN on the proxy
- Internet -> app3.example.com:80 --> nginx_proxy --> NOT OPEN on the proxy
- Internet -> app1.example.com:443 --> nginx_proxy --> docker app1:443
- Internet -> app2.example.com:443 --> nginx_proxy --> docker app2:443
- Internet -> app3.example.com:443 --> nginx_proxy --> docker app2:80  ==> 443 requests forwarded to port 80. 

Can I use VIRTUAL_PORT and VIRTUAL_PROTO on the containers for this usage?

Regards,

Christophe

**Comment by kamermans** [comment]
Hi Christophe,

You do not need to use any directives to do this - it is the default behavior.  Requests on 443 are normally forwarded to port 80 in this scenario.

Thanks,

Steve Kamerman

**Comment by arnaldop** [comment]
Ok, I've had a question for a while and it ties directly to the original question above.

I'm trying to install Bitwarden behind this reverse proxy. Due to the way their install script works, I am limited in how I can configure this project to serve as a reverse proxy for Bitwarden, but I'm trying. Thankfully this project is very flexible.

What I want is this:

`internet --HTTPS(443)--> nginx_proxy --HTTP(non-80 port)--> Bitwarden Nginx`

I would prefer to have SSL all the way through, but currently I can't do that. (Bitwarden expects to get the LE SSL cert directly, but it can't because the reverse proxy is in the middle. But I can't get it with the reverse proxy without Bitwarden running. Chicken and egg problem.)

Regardless, this use case is one I've come across before. I don't necessarily want to configure SSL in every service behind the reverse proxy.

So the question is, what environment variables do I need to set in Bitwarden Nginx in order for the nginx_proxy to receive 443 SSL traffic and forward it to some other non-SSL port?

Currently I have this:
```
      - VIRTUAL_HOST=bitwarden.mydomain.com
      #- VIRTUAL_PROTO=http
      - VIRTUAL_PORT=65080
      #- HTTPS_METHOD=nohttps
      - LETSENCRYPT_HOST=bitwarden.mydomain.com #needed by the letsencrypt companion image
      - LETSENCRYPT_EMAIL=my@email.com #needed by the letsencrypt companion image
```

I have a Docker network that both jwilder-nginx and bitwarden-nginx belong to, so traffic can be forwarded. I also see this in the generated nginx/default.conf:

```
# bitwarden.mydomain.com
upstream bitwarden.mydomain.com {
                                # Cannot connect to network of this container
                                server 127.0.0.1 down;
                                ## Can be connected with "proxynet" network
                # bitwarden-nginx
                        server 172.18.0.9 down;
}
server {
        server_name bitwarden.mydomain.com;
        listen 80 ;
        listen [::]:80 ;
        access_log /var/log/nginx/access.log vhost;
        include /etc/nginx/vhost.d/default;
        location / {
                proxy_pass http://bitwarden.mydomain.com;
        }
}
server {
        server_name bitwarden.mydomain.com;
        listen 443 ssl http2 ;
        listen [::]:443 ssl http2 ;
        access_log /var/log/nginx/access.log vhost;
        return 500;
        ssl_certificate /etc/nginx/certs/default.crt;
        ssl_certificate_key /etc/nginx/certs/default.key;
}
```
It seems that the 2nd `server` block disallows 443 traffic to come directly to it, which is totally fine right now.

And to ensure all external HTTP traffic goes to HTTPS, I have this in additional.conf:

```
server {
        listen 80 default_server;
        listen [::]:80 default_server;
        server_name _;
        return 301 https://$host$request_uri;
}
```

Running `curl -k -v http://localhost:65080` from the Docker host gives me what I am looking for.
Running `curl -k -v https://localhost --header 'Host: bitwarden.mydomain.com'` from the Docker host throws the 500 error.
Running `curl -v -k http://bitwarden.mydomain.com` from another machine on the network gives me a `502 Bad Gateway` with this log message:

```
proxy_1        | nginx.1    | 2018/09/28 21:36:11 [error] 610#610: *695 no live upstreams while connecting to upstream, client: 192.168.1.1, server: bitwarden.mydomain.com, request: "GET / HTTP/1.1", upstream: "http://bitwarden.mydomain.com/", host: "bitwarden.mydomain.com"
```

BTW, I want to keep port 80 open on the reverse proxy so traffic on port 80 can be redirected to port 443.

Any help would be greatly appreciated!
